expression/scalar_function.go

```diff
@@ -291,6 +291,9 @@ func (sf *ScalarFunction) Eval(row chunk.Row) (d types.Datum, err error) {
 
 // EvalInt implements Expression interface.
 func (sf *ScalarFunction) EvalInt(ctx sessionctx.Context, row chunk.Row) (int64, bool, error) {
+	if f, ok := sf.Function.(builtinFuncNew); ok {
+		return f.evalIntWithCtx(ctx, row)
+	}
 	return sf.Function.evalInt(row)
```

Comment:
For using the context in current session. I think of another way:
How about add a new function `setCtx` for struct `builtinFunc`,
and change here as:
```suggestion
	sf.Function.setCtx(ctx)
	return sf.Function.evalInt(row)
```

Reply:
The `setCtx` way is ugly IMHO. @Deardrops 
The real problem is, **the session context is not something belongs to a function**.
As long as function store the `ctx`, the function is invalid when the session context changes.
So the right way to fix the problem is removing `ctx` from the struct, instead of providing `setCtx` as a workaround.